pennylane/tape/qnode.py

```diff
@@ -513,13 +513,23 @@ def construct(self, args, kwargs):
             )
 
     def __call__(self, *args, **kwargs):
+
+        # if shots specified with call, pop it
+        shots = kwargs.pop("shots", None)
+        if shots is not None:
+            original_shots = self.device.shots  # remember device shots
+            self.device.shots = shots  # temporarily change shots
+
         if self.mutable or self.qtape is None:
             # construct the tape
             self.construct(args, kwargs)
 
         # execute the tape
         res = self.qtape.execute(device=self.device)
 
+        if shots is not None:
+            self.device.shots = original_shots
+

```

Comment:
(This doesn't have to be done in this PR, just an idea).

Setting an attribute on an object and then unsetting it after a function call can get messy and confusing, especially when trying to debug applications. The more "functional" way to do it would be to have `execute` always take a `shots` argument explicitly.

```python
shots = kwargs.pop("shots", None) or device.shots
self.execute(device=self.device, shots=shots)
```

The device can then remain totally immutable, and there can never be any accidental state poisoning between runs.

Reply:
I'm very much a fan of this

Reply:
Me too!

Reply:
I checked, and it is a bigger effort. The device class uses its shots argument pretty much as a global argument, i.e. `execute` does not even mention shots in `Device`. So I would suggest we tackle this when we overhaul the Device class one day!

Reply:
Let's put it down for next quarter - with the old core gone, and improvements on the table like moving decompositions to the device and overhauling inv, it feels like the time!